Upgrade Parser to 4.06 AST

anmonteiro opened the pull request:
This PR upgrades the parser to the 4.06 AST version.

This paves the way to solving a number of issues that are currently blocked because we use the 4.04 version of the AST. For example, this PR also fixes:

- #1828 
- #2302 

In addition, this PR introduces parsing / printing of some new syntactic constructs made possible by OCaml 4.06:

- [x] `open` statements in class _expressions_ (e.g. `class x = { open M; as self; };` (https://github.com/ocaml/ocaml/pull/1249)
- [x] `open` statements in class _types_ (e.g. `class type x = { open M; as 'a; };` (also https://github.com/ocaml/ocaml/pull/1249)
- [x] Inherited field in object type (https://github.com/ocaml/ocaml/pull/1118). This is the current syntax for inherited fields in object types:

```rust
type x = {. a: int};
type y = {
  .
  b: string,
  ...x, /* <-- NEW */
};
```

anmonteiro commented on src/reason-parser/reason_parser.mly:
```diff
@@ -1397,7 +1400,6 @@ conflicts.
 %on_error_reduce structure_item let_binding_body
                  as_loc(attribute)+
                  type_longident
-                 attribute+
```
Menhir complains these are unreachable from a toplevel expression. Probably because there's `as_loc(attribute)+` above.

anmonteiro commented:
Tests are now failing on versions of OCaml < 4.06 because I implemented parsing of `Pcl_open` (e.g. `class x = { open M; as self; };`.

Not sure what's the proper way to parse these on older versions of OCaml; or is it even supported?

hcarty commented:
It was only added in 4.06.0. See https://github.com/ocaml/ocaml/blob/trunk/Changes#L801

anmonteiro commented:
@hcarty I'm aware. I'm trying to see if there's a way to run those tests only against 4.06.0.

anmonteiro commented:
I can't find a way to do it with the current setup. In any case, after thinking about it, I don't see the value of exercising the OCaml parser, since it's something we don't control. I'll test `Pcl_open` and the new AST types in `.re` files.

hcarty commented:
@anmonteiro I apologize - I misunderstood the question

anmonteiro commented:
I think this PR is now ready. Code reviews appreciated!

IwanKaramazow commented on src/reason-parser/reason_oprint.ml:
```diff
@@ -444,8 +444,8 @@ and print_simple_out_type ppf =
-        | Ovar_name (id, tyl) ->
-            fprintf ppf "@[%a%a@]" print_typargs tyl print_ident id
+        | Ovar_typ o -> print_out_type ppf o
+            (* fprintf ppf "@[%a%a@]" print_typargs tyl print_ident id *)
```
Commented out line can probably be omitted?

IwanKaramazow commented on src/reason-parser/reason_parser.mly:
```diff
@@ -2001,6 +2003,8 @@ mark_position_cl
   | let_bindings SEMI class_expr_lets_and_rest
     { class_of_let_bindings $1 $3 }
   | object_body { mkclass (Pcl_structure $1) }
+  | LET? OPEN override_flag as_loc(mod_longident) SEMI class_expr_lets_and_rest
```
Is there a reason why we still need to support `let open ...` vs just `open ...` in newly added syntax? I tend to favour just `open` in favour of less grammar complexity. Thoughts?

IwanKaramazow commented on src/reason-parser/reason_parser.mly:
```diff
@@ -2289,17 +2293,24 @@ mark_position_cty
 
+class_sig_body_cty:
+  | class_sig_body { Pcty_signature $1 }
+  | LET? OPEN override_flag as_loc(mod_longident) SEMI as_loc(class_sig_body_cty)
```
Same remark about `let open` as above.

IwanKaramazow commented on src/reason-parser/reason_pprint_ast.ml:
```diff
@@ -2968,7 +2975,7 @@ let printer = object(self:'self)
                 let pcd_name = {
-                  txt = label;
+                  txt = label.txt;
                   loc = pcd_loc;
```
This should probably become `label.loc`? (And finally fixes the comment duplication?)

anmonteiro commented:
Went through the code review items and:

- added typechecked tests for the new AST constructs
- fixed #1828 now that we have locations of each poly variant in the AST (thanks @IwanKaramazow!)

jordwalke commented:
VERY cool.

anmonteiro commented:
superseded by #2487
